File: src/cacheControl.ts
/**
 * cacheControl.ts — Cache-Control directive hygiene audit.
 * T76 cycle 28.
 *
 * Real defect class this catches: Web Cache Deception (Omer
 * Gil, 2017). An attacker tricks an intermediate cache (CDN,
 * reverse proxy, browser cache shared across users on a
 * kiosk) into storing a per-user personalised response by
 * appending a fake static-asset extension to the URL
 * (`/account/foo.css`). If the response carries Set-Cookie
 * but Cache-Control allows shared caching, the next visitor
 * to the cached URL receives the previous user's session +
 * personal data.
 *
 * The defence is a single Cache-Control directive: `no-store`
 * forbids any cache from storing the response. `private`
 * allows browser-only caching but forbids shared/proxy
 * caches — adequate for personalised but non-sensitive
 * content. `public` explicitly invites shared caching and is
 * incompatible with Set-Cookie.
 *
 * Findings:
 *
 *   - cache-control.missing                warn
 *     No Cache-Control header at all. The browser falls back
 *     to RFC 7234 heuristic freshness (typically 10% of the
 *     Last-Modified age). Unpredictable across cache
 *     implementations — be explicit.
 *
 *   - cache-control.public-with-cookie     strict
 *     Response carries Set-Cookie AND Cache-Control includes
 *     `public`. Cache deception risk: an intermediate cache
 *     can store the response keyed by URL alone, then serve
 *     it (with the original Set-Cookie!) to the next visitor.
 *
 *   - cache-control.no-private-with-cookie warn
 *     Response carries Set-Cookie AND Cache-Control omits
 *     both `no-store` and `private`. Less acute than
 *     public-with-cookie but still cacheable by sufficiently
 *     permissive proxies.
 *
 *   - cache-control.invalid                warn
 *     Header value couldn't be parsed into any directive.
 *     Browsers / proxies fall back to no-Cache-Control
 *     semantics — silently disables the operator's intent.
 *
 *   - cache-control.unrealistic-maxage     warn
 *     `max-age` greater than 31536000 (1 year). RFC 7234
 *     section 5.2.1.1 says caches SHOULD treat values
 *     greater than 1 year as 1 year — anything larger is
 *     dead code at best, intent-obscuring at worst.
 *
 *   - cache-control.contradictory          warn
 *     Directive set contains contradictions: both `no-store`
 *     and `max-age` (no-store wins, max-age is dead). Both
 *     `private` and `public` (the spec is ambiguous; most
 *     implementations honour `private`). Both `no-cache` and
 *     `immutable` (immutable means "skip revalidation",
 *     no-cache means "always revalidate" — they cancel).
 *
 * Out of scope:
 *   * Pragma: no-cache (legacy HTTP/1.0 fallback) — separate
 *     concern; if the operator sets BOTH Pragma + Cache-
 *     Control with consistent intent, no defect.
 *   * Expires header — superseded by Cache-Control max-age.
 *     Don't double-flag.
 *   * Per-resource Cache-Control on sub-resources — out of
 *     scope for this top-level detector. Could be a future
 *     per-sub-resource detector once the capture path
 *     established by CORP gets reused.
 *   * Localhost (consistent with the response-header detector
 *     family).
 *
 * Reads from the same `topLevelResponseHeaders` Map as the
 * other 9 response-header detectors. TENTH consumer of the
 * shared capture path. Uses the cycle-24 `responseHeader-
 * Detector` helper.
 */

export interface CacheControlFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface ParsedCacheControl {
  /** All directive names (lowercased), in declaration order. */
  directives: string[];
  /** Lookup of directive name → value (for value-bearing directives). */
  values: Record<string, string>;
  /** True if the header was present but no directive parsed. */
  unparseable: boolean;
}

export interface CacheControlSnapshot {
  pageUrl: string;
  pageIsLocalhost: boolean;
  /** Raw Cache-Control header value, or null. */
  raw: string | null;
  /** Parsed form. Empty if no header. */
  parsed: ParsedCacheControl;
  /** True iff the response also carries Set-Cookie. */
  hasSetCookie: boolean;
}

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return h === 'localhost' || h === '127.0.0.1' || h === '::1' || h.endsWith('.localhost');
  } catch {
    return false;
  }
}

function getHeader(headers: Record<string, string> | undefined, name: string): string | null {
  if (!headers) return null;
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === name) return v;
  }
  return null;
}

function parseCacheControl(raw: string): ParsedCacheControl {
  const directives: string[] = [];
  const values: Record<string, string> = {};
  for (const part of raw.split(',')) {
    const trimmed = part.trim();
    if (!trimmed) continue;
    const eq = trimmed.indexOf('=');
    let name: string;
    let value = '';
    if (eq < 0) {
      name = trimmed.toLowerCase();
    } else {
      name = trimmed.slice(0, eq).trim().toLowerCase();
      value = trimmed.slice(eq + 1).trim();
      // Strip quotes from values (max-age="3600").
      if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
        value = value.slice(1, -1);
      }
    }
    if (!name) continue;
    directives.push(name);
    if (value) values[name] = value;
  }
  const unparseable = raw.trim().length > 0 && directives.length === 0;
  return { directives, values, unparseable };
}

export function buildCacheControlSnapshot(
  pageUrl: string,
  headers: Record<string, string> | undefined,
): CacheControlSnapshot {
  const pageIsLocalhost = isLocalhost(pageUrl);
  const raw = getHeader(headers, 'cache-control');
  const parsed = raw === null
    ? { directives: [], values: {}, unparseable: false }
    : parseCacheControl(raw);
  const hasSetCookie = getHeader(headers, 'set-cookie') !== null;
  return { pageUrl, pageIsLocalhost, raw, parsed, hasSetCookie };
}

const ONE_YEAR_SECONDS = 31536000;

export function detectCacheControlIssues(snap: CacheControlSnapshot): CacheControlFinding[] {
  if (snap.pageIsLocalhost) return [];
  const out: CacheControlFinding[] = [];

  if (snap.raw === null) {
    out.push({
      severity: 'warn',
      kind: 'cache-control.missing',
      detail: `No Cache-Control header. Browsers and intermediate caches fall back to RFC 7234 heuristic freshness (typically 10% of the Last-Modified age) — unpredictable across implementations. Be explicit: 'Cache-Control: no-store' for sensitive pages, 'private, max-age=<seconds>' for personalised but cacheable, 'public, max-age=<seconds>, immutable' for static assets.`,
      evidence: { hasSetCookie: snap.hasSetCookie },
    });
    return out;
  }

  if (snap.parsed.unparseable) {
    out.push({
      severity: 'warn',
      kind: 'cache-control.invalid',
      detail: `Cache-Control header is set but couldn't be parsed into any directive. Browsers and proxies fall back to no-Cache-Control semantics, silently disabling the operator's intent. Header value: '${snap.raw.slice(0, 200)}'.`,
      evidence: { raw: snap.raw.slice(0, 200) },
    });
    return out;
  }

  const dset = new Set(snap.parsed.directives);

  // --- Set-Cookie + caching combinations ---
  if (snap.hasSetCookie && dset.has('public')) {
    out.push({
      severity: 'strict',
      kind: 'cache-control.public-with-cookie',
      detail: `Response sets a Set-Cookie header AND Cache-Control includes 'public'. Web Cache Deception risk: an intermediate cache (CDN, reverse proxy, kiosk browser) can store the response keyed by URL, then serve it WITH THE ORIGINAL Set-Cookie to the next visitor. Replace 'public' with 'no-store' (sensitive) or 'private' (personalised, browser-only).`,
      evidence: { directives: snap.parsed.directives },
    });
  } else if (snap.hasSetCookie && !dset.has('no-store') && !dset.has('private')) {
    out.push({
      severity: 'warn',
      kind: 'cache-control.no-private-with-cookie',
      detail: `Response sets a Set-Cookie header but Cache-Control omits both 'no-store' and 'private'. Sufficiently permissive proxies (some CDNs, corporate caches) may store and serve the response cross-user. Add 'private' (browser-only caching) or 'no-store' (no caching at all).`,
      evidence: { directives: snap.parsed.directives },
    });
  }

  // --- max-age sanity ---
  const maxAge = snap.parsed.values['max-age'];
  if (maxAge !== undefined) {
    const n = Number(maxAge);
    if (!Number.isFinite(n) || n < 0) {
      out.push({
        severity: 'warn',
        kind: 'cache-control.invalid',
        detail: `Cache-Control max-age value '${maxAge}' is not a non-negative integer. Browsers / proxies typically ignore.`,
        evidence: { maxAge },
      });
    } else if (n > ONE_YEAR_SECONDS) {
      out.push({
        severity: 'warn',
        kind: 'cache-control.unrealistic-maxage',
        detail: `Cache-Control max-age=${n} exceeds 1 year (31536000s). RFC 7234 §5.2.1.1: caches SHOULD treat values greater than 1 year as 1 year — anything larger is dead code at best.`,
        evidence: { maxAge: n },
      });
    }
  }

  // --- Contradictory directive combinations ---
  const contradictions: string[] = [];
  if (dset.has('no-store') && (dset.has('max-age') || dset.has('s-maxage'))) {
    contradictions.push("'no-store' + 'max-age': no-store wins, max-age is dead");
  }
  if (dset.has('public') && dset.has('private')) {
    contradictions.push("'public' + 'private': spec ambiguous, most implementations honour 'private'");
  }
  if (dset.has('no-cache') && dset.has('immutable')) {
    contradictions.push("'no-cache' + 'immutable': cancel each other (revalidate-always vs skip-revalidate)");
  }
  if (dset.has('no-store') && dset.has('immutable')) {
    contradictions.push("'no-store' + 'immutable': no-store forbids any cache, immutable assumes a cache");
  }
  if (contradictions.length > 0) {
    out.push({
      severity: 'warn',
      kind: 'cache-control.contradictory',
      detail: `Cache-Control contains contradictory directives. ${contradictions.join('. ')}. Pick one intent and stick with it.`,
      evidence: { directives: snap.parsed.directives, contradictions },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/cache_control.rs
//! `cache_control` — Cache-Control directive hygiene detector.
//!
//! Mirror of `src/cacheControl.ts`. Catches Web Cache Deception
//! (Omer Gil, 2017) + sloppy directive hygiene. Findings:
//!
//!   * `cache-control.missing`                 warn
//!   * `cache-control.public-with-cookie`      strict
//!   * `cache-control.no-private-with-cookie`  warn
//!   * `cache-control.invalid`                 warn
//!   * `cache-control.unrealistic-maxage`      warn  (>1y)
//!   * `cache-control.contradictory`           warn
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::url_helpers::is_localhost;
use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

const ONE_YEAR_SECONDS: u64 = 31_536_000;

/// Parsed form of the Cache-Control header value.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedCacheControl {
    /// All directive names (lowercased), in declaration order.
    pub directives: Vec<String>,
    /// Lookup of directive name → value (for value-bearing forms).
    pub values: HashMap<String, String>,
    /// True iff the header was present but no directive parsed.
    pub unparseable: bool,
}

/// Captured page state.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CacheControlSnapshot {
    /// Page URL (used for localhost exemption).
    pub page_url: String,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// Raw header value, `None` if absent.
    pub raw: Option<String>,
    /// Parsed form. Default-empty if `raw` is `None`.
    pub parsed: ParsedCacheControl,
    /// True iff the response also carries Set-Cookie.
    pub has_set_cookie: bool,
}

fn lookup_header<I, K, V>(headers: I, name_lower: &str) -> Option<String>
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    for (k, v) in headers {
        if k.as_ref().to_ascii_lowercase() == name_lower {
            return Some(v.as_ref().to_owned());
        }
    }
    None
}

fn parse_cache_control(raw: &str) -> ParsedCacheControl {
    let mut directives = Vec::new();
    let mut values: HashMap<String, String> = HashMap::new();
    for part in raw.split(',') {
        let trimmed = part.trim();
        if trimmed.is_empty() {
            continue;
        }
        let (name, value) = match trimmed.find('=') {
            None => (trimmed.to_ascii_lowercase(), String::new()),
            Some(eq) => {
                let name = trimmed[..eq].trim().to_ascii_lowercase();
                let mut value = trimmed[eq + 1..].trim().to_owned();
                if (value.starts_with('"') && value.ends_with('"') && value.len() >= 2)
                    || (value.starts_with('\'') && value.ends_with('\'') && value.len() >= 2)
                {
                    value = value[1..value.len() - 1].to_owned();
                }
                (name, value)
            }
        };
        if name.is_empty() {
            continue;
        }
        directives.push(name.clone());
        if !value.is_empty() {
            values.insert(name, value);
        }
    }
    let unparseable = !raw.trim().is_empty() && directives.is_empty();
    ParsedCacheControl { directives, values, unparseable }
}

/// Build a snapshot from a page URL + a header map. The map must
/// expose every relevant header — Cache-Control + Set-Cookie at
/// minimum. Headers are looked up case-insensitively.
pub fn build_cache_control_snapshot<I, K, V>(
    page_url: &str,
    headers: I,
) -> CacheControlSnapshot
where
    I: IntoIterator<Item = (K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    // Collect headers once because the iterator is consumed.
    let collected: Vec<(String, String)> = headers
        .into_iter()
        .map(|(k, v)| (k.as_ref().to_owned(), v.as_ref().to_owned()))
        .collect();
    let raw = lookup_header(collected.iter().map(|(k, v)| (k.as_str(), v.as_str())), "cache-control");
    let parsed = match &raw {
        None => ParsedCacheControl::default(),
        Some(r) => parse_cache_control(r),
    };
    let has_set_cookie = lookup_header(
        collected.iter().map(|(k, v)| (k.as_str(), v.as_str())),
        "set-cookie",
    )
    .is_some();
    CacheControlSnapshot {
        page_url: page_url.to_owned(),
        page_is_localhost: is_localhost(page_url),
        raw,
        parsed,
        has_set_cookie,
    }
}

/// Run the detector. Returns zero findings on localhost.
pub fn detect_cache_control_issues(snap: &CacheControlSnapshot) -> Vec<AxisFinding> {
    if snap.page_is_localhost {
        return Vec::new();
    }
    let mut out = Vec::new();

    let Some(raw) = snap.raw.as_deref() else {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "cache-control.missing".into(),
            detail: format!(
                "No Cache-Control header. Browsers and intermediate caches fall back to \
                 RFC 7234 heuristic freshness (typically 10% of Last-Modified age) — \
                 unpredictable across implementations. Be explicit: 'Cache-Control: \
                 no-store' for sensitive pages, 'private, max-age=<seconds>' for \
                 personalised but cacheable, 'public, max-age=<seconds>, immutable' \
                 for static assets. (hasSetCookie={})",
                snap.has_set_cookie
            ),
        });
        return out;
    };

    if snap.parsed.unparseable {
        let preview = &raw[..raw.len().min(200)];
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "cache-control.invalid".into(),
            detail: format!(
                "Cache-Control header is set but couldn't be parsed into any directive. \
                 Browsers and proxies fall back to no-Cache-Control semantics, silently \
                 disabling the operator's intent. Header value: '{preview}'."
            ),
        });
        return out;
    }

    let dset: std::collections::HashSet<&str> =
        snap.parsed.directives.iter().map(|s| s.as_str()).collect();

    if snap.has_set_cookie && dset.contains("public") {
        out.push(AxisFinding {
            severity: AxisSeverity::Strict,
            kind: "cache-control.public-with-cookie".into(),
            detail: format!(
                "Response sets a Set-Cookie header AND Cache-Control includes 'public'. \
                 Web Cache Deception risk: an intermediate cache (CDN, reverse proxy, \
                 kiosk browser) can store the response keyed by URL, then serve it WITH \
                 THE ORIGINAL Set-Cookie to the next visitor. Replace 'public' with \
                 'no-store' (sensitive) or 'private' (personalised, browser-only). \
                 (directives={:?})",
                snap.parsed.directives
            ),
        });
    } else if snap.has_set_cookie
        && !dset.contains("no-store")
        && !dset.contains("private")
    {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "cache-control.no-private-with-cookie".into(),
            detail: format!(
                "Response sets a Set-Cookie header but Cache-Control omits both \
                 'no-store' and 'private'. Sufficiently permissive proxies (some CDNs, \
                 corporate caches) may store and serve the response cross-user. \
                 Add 'private' (browser-only caching) or 'no-store' (no caching at all). \
                 (directives={:?})",
                snap.parsed.directives
            ),
        });
    }

    if let Some(max_age_str) = snap.parsed.values.get("max-age") {
        match max_age_str.parse::<i64>() {
            Err(_) => {
                out.push(AxisFinding {
                    severity: AxisSeverity::Warn,
                    kind: "cache-control.invalid".into(),
                    detail: format!(
                        "Cache-Control max-age value '{max_age_str}' is not a \
                         non-negative integer. Browsers / proxies typically ignore."
                    ),
                });
            }
            Ok(n) if n < 0 => {
                out.push(AxisFinding {
                    severity: AxisSeverity::Warn,
                    kind: "cache-control.invalid".into(),
                    detail: format!(
                        "Cache-Control max-age value '{max_age_str}' is negative. \
                         Browsers / proxies typically ignore."
                    ),
                });
            }
            Ok(n) if (n as u64) > ONE_YEAR_SECONDS => {
                out.push(AxisFinding {
                    severity: AxisSeverity::Warn,
                    kind: "cache-control.unrealistic-maxage".into(),
                    detail: format!(
                        "Cache-Control max-age={n} exceeds 1 year (31536000s). \
                         RFC 7234 §5.2.1.1: caches SHOULD treat values greater than \
                         1 year as 1 year — anything larger is dead code at best."
                    ),
                });
            }
            _ => {}
        }
    }

    let mut contradictions: Vec<String> = Vec::new();
    if dset.contains("no-store")
        && (dset.contains("max-age") || dset.contains("s-maxage"))
    {
        contradictions
            .push("'no-store' + 'max-age': no-store wins, max-age is dead".into());
    }
    if dset.contains("public") && dset.contains("private") {
        contradictions.push(
            "'public' + 'private': spec ambiguous, most implementations honour 'private'"
                .into(),
        );
    }
    if dset.contains("no-cache") && dset.contains("immutable") {
        contradictions.push(
            "'no-cache' + 'immutable': cancel each other (revalidate-always vs skip-revalidate)"
                .into(),
        );
    }
    if dset.contains("no-store") && dset.contains("immutable") {
        contradictions.push(
            "'no-store' + 'immutable': no-store forbids any cache, immutable assumes one"
                .into(),
        );
    }
    if !contradictions.is_empty() {
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "cache-control.contradictory".into(),
            detail: format!(
                "Cache-Control contains contradictory directives. {}. Pick one intent \
                 and stick with it. (directives={:?})",
                contradictions.join(". "),
                snap.parsed.directives
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn snap(url: &str, cc: Option<&str>, set_cookie: bool) -> CacheControlSnapshot {
        let mut headers: Vec<(&str, &str)> = Vec::new();
        if let Some(v) = cc {
            headers.push(("cache-control", v));
        }
        if set_cookie {
            headers.push(("set-cookie", "session=abc; Path=/"));
        }
        build_cache_control_snapshot(url, headers)
    }

    #[test]
    fn localhost_exempt() {
        assert!(detect_cache_control_issues(&snap("https://localhost/", None, true)).is_empty());
    }

    #[test]
    fn missing_header_warn() {
        let f = detect_cache_control_issues(&snap("https://example.com/", None, false));
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "cache-control.missing");
    }

    #[test]
    fn public_with_set_cookie_strict() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, max-age=600"),
            true,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.public-with-cookie"
            && x.severity == AxisSeverity::Strict));
    }

    #[test]
    fn cookie_without_private_or_no_store_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("max-age=600"),
            true,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.no-private-with-cookie"));
    }

    #[test]
    fn private_with_set_cookie_clean() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("private, max-age=0"),
            true,
        ));
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn no_store_with_set_cookie_clean() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("no-store"),
            true,
        ));
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn unrealistic_maxage_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, max-age=99999999"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.unrealistic-maxage"));
    }

    #[test]
    fn invalid_maxage_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, max-age=banana"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.invalid"));
    }

    #[test]
    fn negative_maxage_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, max-age=-1"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.invalid"));
    }

    #[test]
    fn contradictory_no_store_max_age_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("no-store, max-age=600"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.contradictory"));
    }

    #[test]
    fn contradictory_public_private_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, private"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.contradictory"));
    }

    #[test]
    fn contradictory_no_cache_immutable_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("no-cache, immutable, max-age=600"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.contradictory"));
    }

    #[test]
    fn contradictory_no_store_immutable_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("no-store, immutable"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.contradictory"));
    }

    #[test]
    fn unparseable_value_invalid_warn() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some(",,,,,"),
            false,
        ));
        assert!(f.iter().any(|x| x.kind == "cache-control.invalid"));
    }

    #[test]
    fn case_insensitive_directives() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("PUBLIC, MAX-AGE=600"),
            true,
        ));
        // Public + cookie still triggers strict — case-insensitive parse.
        assert!(f.iter().any(|x| x.kind == "cache-control.public-with-cookie"));
    }

    #[test]
    fn quoted_max_age_accepted() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, max-age=\"3600\""),
            false,
        ));
        // No findings — value parses, no-cookie scenario, sensible TTL.
        assert!(f.is_empty(), "got {f:?}");
    }

    #[test]
    fn header_lookup_case_insensitive() {
        let s = build_cache_control_snapshot(
            "https://example.com/",
            [("Cache-Control", "private, max-age=600")],
        );
        assert!(detect_cache_control_issues(&s).is_empty());
    }

    #[test]
    fn one_year_boundary_is_acceptable() {
        let f = detect_cache_control_issues(&snap(
            "https://example.com/",
            Some("public, max-age=31536000"),
            false,
        ));
        assert!(f.is_empty(), "1y exact is OK; got {f:?}");
    }
}
